### PR TITLE
Reorganize application files and refactor gateways

### PR DESCRIPTION
**Pull request by aabadie**
This PR fixes several inconsistencies and bad naming in all applications (broker, dashboard, gateways):
* `application.py` and specific application modules (gateway.py, broker.py, etc) have been exchanged, making the module names more consistent
* The gateway, broker and dashboard classes have also been renamed

The last big change removes code duplication in gateways by introducing a base gateway object.

### CODE REVIEW

**Comment by aabadie** [comment]
Maybe @bergzand can give some help to review the changes here.

**Comment by bergzand** [comment]
That's a lot of refactoring! I should have some time this weekend for a review.

File: pyaiot/gateway/common/common.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+from tornado import web, gen
+from tornado.websocket import websocket_connect
+
+from pyaiot.common.auth import auth_token
+
+logger = logging.getLogger("pyaiot.gw.common")
+
+
+class GatewayBase(web.Application):
+    """Tornado based gateway application for CoAP nodes on a network."""
```
Comment (bergzand) [inline pyaiot/gateway/common/common.py]: Maybe reword to "Base class for Tornado based gateway applications" as this is not a CoAP specific base class

File: pyaiot/gateway/common/common.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        # Create connection to broker
+        self.create_broker_connection(
+            "ws://{}:{}/gw".format(options.broker_host, options.broker_port))
+
+        super().__init__(handlers, **settings)
+        logger.debug('Base Gateway application started')
+
+
+    def setup_nodes_controller(self):
+        """Instanciate and configure an application nodes controller."""
```
Comment (bergzand) [inline pyaiot/gateway/common/common.py]: Small typo: "Instantiate"

File: pyaiot/gateway/mqtt/gateway.py [hunk 3]
```diff
@@ -27,81 +27,34 @@
...
-    try:
-        parse_command_line()
-    except SyntaxError as e:
-        logger.error("Invalid config file: {}".format(e))
-        return
-    except FileNotFoundError as e:
-        logger.error("Config file not found: {}".format(e))
-        return
+    def setup_nodes_controller(self):
+        """Instanciate and configure an MQTT nodes controller."""
```
Comment (bergzand) [inline pyaiot/gateway/mqtt/gateway.py]: Small typo: "Instantiate"

File: pyaiot/gateway/common/common.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        if options.debug:
+            logger.setLevel(logging.DEBUG)
+
+        self.broker = None
+        self.keys = keys
+        handlers = []
+        settings = {'debug': True}
+
+        # Setup nodes controller
+        self._nodes_controller = self.setup_nodes_controller()
```
Comment (bergzand) [inline pyaiot/gateway/common/common.py:56]: Not completely sure on this, but maybe add a dummy `setup_nodes_controller()` to this class

File: pyaiot/gateway/coap/gateway.py [hunk 5]
```diff
@@ -27,80 +27,31 @@
...
 
-    start_application(CoapGatewayApplication(keys, options=options),
-                      port=options.coap_port, close_client=True)
+    def __init__(self, keys, options):
+        self.options = options
+        super().__init__(keys, options)
+        logger.info('CoAP gateway application started')
 
+    def setup_nodes_controller(self):
+        """Instanciate and configure a CoAP nodes controller."""
```
Comment (bergzand) [inline pyaiot/gateway/coap/gateway.py]: Small typo: "Instantiate"

File: pyaiot/gateway/ws/gateway.py [hunk 6]
```diff
@@ -27,73 +27,112 @@
...
+            (r"/node", WebsocketNodeHandler),
+        ]
+
+        super().__init__(keys, options)
+
+        logger.info('WS gateway started, listening on port {}'
+                    .format(options.gateway_port))
+
+    def setup_nodes_controller(self):
+        """Instanciate and configure a websocket nodes controller."""
```
Comment (bergzand) [inline pyaiot/gateway/ws/gateway.py]: Small typo: "Instantiate"

**Review by bergzand — CHANGES_REQUESTED** [review]
Nothing big here, a few minor comments.

**Comment by bergzand** [comment]
Looks good overall, I didn't test it yet, but the overall changes look good to me. Good work with the refactor to a single base gateway class 👍

**Comment by bergzand** [comment]
There are also a few pep8 issues. Do you want to enforce pep8 on PR changes?

File: pyaiot/gateway/common/common.py (hunk 4 above)
Comment (aabadie) [reply pyaiot/gateway/common/common.py:56]: You hit the weak point of this PR :)
With this version, it can happen that a concrete gateway doesn't implement this method and the `_nodes_controller` could be `None`

I changed that to use multiple inheritance using `ABCMeta` and setting the `setup_nodes_controller` as pure virtual. Then all concrete gateway implementations have to implement this method. I tested it and it works well: an error message is displayed when the method is not implemented.
We could go even further by applying a similar design to the node controllers: a base class with abstract methods and concrete classes implementing them.

**Comment by aabadie** [comment]
@bergzand, Thanks for reviewing this. I addressed your comments.

> There are also a few pep8 issues. Do you want to enforce pep8 on PR changes?

Yes, it's always a good thing ! This can be done very easily with pytest. Do you propose yourself for a PR ?

File: pyaiot/gateway/common/common.py (hunk 4 above)
Comment (bergzand) [reply pyaiot/gateway/common/common.py:56]: Nice, this is exactly what I hoped to achieve with this comment.

**Comment by bergzand** [comment]
> Yes, it's always a good thing ! This can be done very easily with pytest. Do you propose yourself for a PR ?

Yes, I can get a PR ready with pep8 fixes and an additional test this week or so.

Tested a bit, looks good.
Ack from me for this PR

**Comment by aabadie** [comment]
> Ack from me for this PR

Thanks, then I'll merge !
